crmsh/join_lock.py

```diff
@@ -0,0 +1,167 @@
+
+
+class JoinLock(object):
+    """
+    Class to manage lock for multiple nodes join in parallel
+    """
+
+    JOIN_LOCK_DIR = "/usr/share/crmsh/.join_lock_directory"
+    JOIN_LOCK_FILE = "{}/join_lock_file".format(JOIN_LOCK_DIR)
+    TAIL_CMD = "tail -1 {}".format(JOIN_LOCK_FILE)
+    MKDIR_CMD = "mkdir {}".format(JOIN_LOCK_DIR)
+    RM_CMD = "rm -rf {}".format(JOIN_LOCK_DIR)
+    UPDATE_CMD = "echo {} `date +%s` >> {}".format(utils.this_node(), JOIN_LOCK_FILE)
+    DETECT_INTERVAL = 15
+    UPDATE_INTERVAL = 10
+    SSH_OPTION = "-o ConnectTimeout={} -o StrictHostKeyChecking=no".format(UPDATE_INTERVAL)
+
+    def __init__(self, target_node):
+        """
+        Init function
+        """
+        self.thread_stop_event = threading.Event()
+        self.target_node = target_node
+
+    def _run(self, cmd):
+        """
+        Run command on target node, consider specific exceptions
+        """
+        cmd_with_ssh = "ssh {} root@{} \"{}\"".format(self.SSH_OPTION, self.target_node, cmd)
+        rc, out, err = utils.get_stdout_stderr(cmd_with_ssh)
+        if rc != 0 and err:
+            if re.search("tail: cannot open .*: No such file or directory", err):
+                raise FileNotFoundError(err)
+            elif re.search("mkdir: cannot create directory .*: File exists", err):
+                raise FileExistsError(err)
+            elif re.search("ssh: connect to .*: Connection timed out", err):
+                raise TimeoutError(err)
+            else:
+                raise RuntimeError(err)
+        return rc, out
+
+    def _get_last_timestamp(self):
+        """
+        Tail the last line in lock file
+        """
+        result = {}
+        rc, out = self._run(self.TAIL_CMD)
+        if rc == 0 and out:
+            try:
+                locked_node, timestamp = out.split()
+            except ValueError:
+                raise ValueError("Invalid format for {}(should be \"node timestamp\")".format(self.JOIN_LOCK_FILE))
+            else:
+                result['node'] = locked_node
+                result['timestamp'] = timestamp
+        return result
+
+    def _remove_lock_dir(self):
+        """
+        Remove lock directory
+        """
+        self._run(self.RM_CMD)
+
+    def _create_lock_dir(self):
+        """
+        Create lock directory, mkdir command was atomic
+        """
+        try:
+            self._run(self.MKDIR_CMD)
+        except FileExistsError:
+            return False
+        return True
+
+    def _lock_or_wait(self):
+        """
+        Wait 15 seconds while detect target node has lock directory
+        """
+        file_not_found_once = False
+        warned_once = False
+        origin_timestamp = None
+
+        while not self._create_lock_dir():
+
+            try:
+                result = self._get_last_timestamp()
+            # consider lock file might not be created yet
+            except FileNotFoundError:
```

Comment:
This looks strange. Exceptions used as normal code flow

Reply:
yep https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why